### PR TITLE
[lwIP/TCPIP] CORE-13098: Upgrade lwIP from 1.4.1 to 2.1.3.

### PR DESCRIPTION
**Pull request by GGLinnk**
Will be refactored to comply with #5025 

## Purpose
Upgrade lwIP lib from 1.4.1 to 2.1.3

JIRA issue: [CORE-13098](https://jira.reactos.org/browse/CORE-13098)

## Proposed changes
- Moved ROS functions from lwIP to it's dedicated lib (lwIP_GLUE).
- Upgraded lwIP from 1.4.1 to 2.1.3.
- Fixed BSOD relative to flag link_up not set. (Thanks to Simon Goldschmidt)
- Added a netif state debug message.

## Related Pull Request
- https://github.com/reactos/reactos/pull/4712
- https://github.com/reactos/reactos/pull/2001

## Related Fixed Issues
- https://jira.reactos.org/browse/CORE-12024

## TODO
- [x] Split initial commit in two piece to improve review process (https://github.com/reactos/reactos/pull/4999#pullrequestreview-1252625672)
- [x] Investigate and fix ``0x0000007E`` BSOD (https://github.com/reactos/reactos/pull/4999#issuecomment-1386021889)
- [x] Investigate and fix "tcp not closed properly" regression (https://github.com/reactos/reactos/pull/4999#issuecomment-1399258969)
- [ ] See with the developers if I should remove the unused applications/code from lwip or if I should keep them?
- [x] Search for further bugs and crashes

### CODE REVIEW

**Comment by KRosUser2** [comment]
updating previous build. Launching K-Meleon 75.1 : BSOD 0x7e

(/ntoskrnl/ps/thread.c:119) PS: Unhandled Kernel Mode Exception Pointers = 0xF6E94670
(/ntoskrnl/ps/thread.c:126) Code c0000005 Addr F6EB3F7E Info0 00000000 Info1 0044007D Info2 00000000 Info3 00000400

*** Fatal System Error: 0x0000007e
                       (0xC0000005,0xF6EB3F7E,0xF6E94B18,0xF6E9481C)


Entered debugger on embedded INT3 at 0x0008:0x8059D87C.
kdb:> bt
Eip:
<ntoskrnl.exe:19d87c (sdk/lib/rtl/i386/debug_asm.S:56 (RtlpBreakWithStatusInstruction))>
Frames:
<ntoskrnl.exe:94f91 (ntoskrnl/ke/bug.c:1066 (KeBugCheckWithTf))>
<ntoskrnl.exe:954fb (ntoskrnl/ke/bug.c:1413 (KeBugCheckEx))>
<ntoskrnl.exe:139649 (ntoskrnl/ps/thread.c:129 (PspUnhandledExceptionInSystemThread))>
<ntoskrnl.exe:13969e (ntoskrnl/ps/thread.c:159 (_SEH3$_FilterFunction.47176))>
<ntoskrnl.exe:177653 (sdk/lib/pseh/i386/pseh3.c:80 (_SEH3$_GetFilterResult))>
<ntoskrnl.exe:17776d (sdk/lib/pseh/i386/pseh3.c:299 (_SEH3$_except_handler))>
<ntoskrnl.exe:19d9d7 (sdk/lib/rtl/i386/except_asm.s:185 (RtlpExecuteHandler2))>
<ntoskrnl.exe:19d9ab (sdk/lib/rtl/i386/except_asm.s:151 (RtlpExecuteHandlerForUnwind))>
<ntoskrnl.exe:15a732 (ntoskrnl/ke/i386/exp.c:872 (KiDispatchException))>
<ntoskrnl.exe:15ac52 (ntoskrnl/ke/i386/exp.c:1081 (KiDispatchExceptionFromTrapFrame))>
<ntoskrnl.exe:15fa9d (ntoskrnl/include/internal/i386/ke.h:770 (KiTrap0EHandler))>
<ntoskrnl.exe:36ae (:0 (KiTrap0E))>
<tcpip.sys:1ef79 (sdk/lib/drivers/ip/lwip/src/core/tcp.c:322 (tcp_backlog_accepted))>
<tcpip.sys:2b512 (sdk/lib/drivers/ip/lwip_glue/rostcp.c:794 (LibTCPAccept))>
<tcpip.sys:1daca (sdk/lib/drivers/ip/transport/tcp/event.c:275 (TCPAcceptEventHandler))>
<tcpip.sys:2a1fd (sdk/lib/drivers/ip/lwip_glue/rostcp.c:280 (InternalAcceptEventHandler))>
<tcpip.sys:22ef3 (sdk/lib/drivers/ip/lwip/src/core/tcp_in.c:941 (tcp_input))>
<tcpip.sys:26d0b (sdk/lib/drivers/ip/lwip/src/core/ipv4/ip4.c:708 (ip4_input))>
<tcpip.sys:2758d (sdk/lib/drivers/ip/lwip/src/api/tcpip.c:173 (tcpip_thread))>
<tcpip.sys:2b5fe (sdk/lib/drivers/ip/lwip_glue/sys_arch.c:285 (LwipThreadMain))>
<ntoskrnl.exe:13954a (ntoskrnl/ps/thread.c:156 (PspSystemThreadStartup))>
<ntoskrnl.exe:15c149 (ntoskrnl/ke/i386/thrdini.c:78 (KiThreadStartup))>
<ntoskrnl.exe:13951d (ntoskrnl/ps/thread.c:63 (PspUserThreadStartup))>
<ec835356>

**Comment by hope2586** [comment]
The mentioned crash only happens when it is compiled with GCC. Compiling it with MSVC doesn't result in the crash.

**Review by Extravert-ir — CHANGES_REQUESTED** [review]
Please divide the library update itself and changes to glue code. It is impossible to review like that

**Comment by GGLinnk** [comment]
Should be easier to review.
I've "injected" an intermediary non-compilable commit that will improve review process.

**Comment by GGLinnk** [comment]
@KRosUser2 : https://github.com/reactos/reactos/pull/4999#issuecomment-1386021889 fixed
VirtualBox_ReactOS - beta_20_01_2023_17_33_47
Please continue testing this PR.

**Comment by KRosUser2** [comment]
Web browseing with K-Meleon OK.

SUMo : One regression at the end of a "check" process : 
image

**Review by ghost — CHANGES_REQUESTED** [review]

File: sdk/lib/drivers/ip/lwip_glue/rostcp.c [hunk 1]
```diff
@@ -840,11 +790,12 @@ LibTCPAccept(PTCP_PCB pcb, struct tcp_pcb *listen_pcb, void *arg)
     tcp_err(pcb, InternalErrorEventHandler);
     tcp_arg(pcb, arg);
 
-    tcp_accepted(listen_pcb);
+    //tcp_accepted(listen_pcb); /* DEPRECATED */
+    tcp_backlog_accepted(listen_pcb);
```
[the remark this answers is not in the document]
Comment (GGLinnk) [reply sdk/lib/drivers/ip/lwip_glue/rostcp.c]: This already have been removed.

**Comment by KRosUser2** [comment]
SUMo : One regression at the end of a 
image
"check" process :

was formerly patched by @sdever in ReactOS lwIP code : https://jira.reactos.org/browse/CORE-16868 https://github.com/reactos/reactos/pull/2582

File: sdk/lib/drivers/ip/lwip/src/core/tcp.c [hunk 2]
```diff
@@ -0,0 +1,2690 @@
...
+ *
+ * @param pcb the connection pcb which is now fully accepted (or closed/aborted)
+ */
+void
+tcp_backlog_accepted(struct tcp_pcb *pcb)
+{
+  LWIP_ASSERT("pcb != NULL", pcb != NULL);
+  LWIP_ASSERT_CORE_LOCKED();
+  if ((pcb->flags & TF_BACKLOGPEND) != 0) {
+    if (pcb->listener != NULL) {
```
[the remark this answers is not in the document]
Comment (GGLinnk) [reply sdk/lib/drivers/ip/lwip/src/core/tcp.c:322]: Outdated. Relative to https://github.com/reactos/reactos/pull/4999#discussion_r1072791825
Resolved

**Comment by KRosUser2** [comment]
Tested OK with
- K-Meleon Web browsing (HTTP)
- K-Meleon FTP
- Filezilla client FTP
- SUMo HTTP

Well done !

**Review by tkreuzer — CHANGES_REQUESTED** [review]
I don't know what you did with the files but seemingly you fucked up the encoding. This makes reviewing impossible. Please make sure you do not change the encoding of any file in any commit.

**Comment by GGLinnk** [comment]
> I don't know what you did with the files but seemingly you fucked up the encoding. This makes reviewing impossible. Please make sure you do not change the encoding of any file in any commit.

What is the encoding initially used by these files?
You you like to be more precise on the issue ?
Can you give me a specific exemple ? (File, line, anything)

**Comment by KRosUser2** [comment]
this PR will fix this BSOD : https://jira.reactos.org/browse/CORE-12024?focusedCommentId=138138&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-138138

**Review by Extravert-ir — CHANGES_REQUESTED** [review]
I looked at these changes and decided to rearrange how `tcpip.sys` libraries are organized. See PR #5025.
Please, rebase this PR ontop of that, and that should be good to go

**Comment by GGLinnk** [comment]
I will close it for now. I don't have time to rebase it for a while. Probably gonna work on it later tho.

**Pull request closed without merging**